ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Embed/Nary.hs

```diff
@@ -185,51 +192,70 @@ instance Inject ExtLedgerState where
 -- problematic, but extending 'ledgerViewForecastAt' is a lot more subtle; see
 -- @forecastNotFinal@.
 injectInitialExtLedgerState ::
-     forall x xs. CanHardFork (x ': xs)
+     forall x xs. (CanHardFork (x ': xs), Untick (Ext.ExtLedgerState (HardForkBlock (x ': xs))))
   => TopLevelConfig (HardForkBlock (x ': xs))
-  -> ExtLedgerState x
-  -> ExtLedgerState (HardForkBlock (x ': xs))
+  -> Ext.ExtLedgerState x
+  -> Ext.ExtLedgerState (HardForkBlock (x ': xs))
 injectInitialExtLedgerState cfg extLedgerState0 =
-    ExtLedgerState {
-        ledgerState = targetEraLedgerState
-      , headerState = targetEraHeaderState
+    go Ext.ExtLedgerState {
+        ledgerState =
+            HardForkLedgerState
+          $ initHardForkState
+          $ Ext.ledgerState extLedgerState0
+      , headerState =
+            genesisHeaderState
+          $ initHardForkState
+          $ WrapChainDepState
+          $ headerStateChainDep
+          $ Ext.headerState extLedgerState0
       }
   where
-    cfgs :: NP TopLevelConfig (x ': xs)
-    cfgs =
-        distribTopLevelConfig
-          (State.epochInfoLedger
-             (configLedger cfg)
-             (hardForkLedgerStatePerEra targetEraLedgerState))
-          cfg
-
-    targetEraLedgerState :: LedgerState (HardForkBlock (x ': xs))
-    targetEraLedgerState =
-        HardForkLedgerState $
-          -- We can immediately extend it to the right slot, executing any
-          -- scheduled hard forks in the first slot
-          State.extendToSlot
-            (configLedger cfg)
-            (SlotNo 0)
-            (initHardForkState (ledgerState extLedgerState0))
-
-    firstEraChainDepState :: HardForkChainDepState (x ': xs)
-    firstEraChainDepState =
-        initHardForkState $
-          WrapChainDepState $
-            headerStateChainDep $
-              headerState extLedgerState0
-
-    targetEraChainDepState :: HardForkChainDepState (x ': xs)
-    targetEraChainDepState =
-        -- Align the 'ChainDepState' with the ledger state of the target era.
-        State.align
-          (InPairs.requiringBoth
-            (hmap (WrapConsensusConfig . configConsensus) cfgs)
-            (translateChainDepState hardForkEraTranslation))
-          (hpure (fn_2 (\_ st -> st)))
-          (hardForkLedgerStatePerEra targetEraLedgerState)
-          firstEraChainDepState
-
-    targetEraHeaderState :: HeaderState (HardForkBlock (x ': xs))
-    targetEraHeaderState = genesisHeaderState targetEraChainDepState
+    go :: Ext.ExtLedgerState (HardForkBlock (x ': xs)) -> Ext.ExtLedgerState (HardForkBlock (x ': xs))
```

Comment:
I think this loop works as-is. It relies on unticking, which is worrying, but relatively harmless, because of the isolation via `LockBox`.

The only rough alternative I've imagined that would likely be preferable is an `NP (ExtLedgerState ByronBlock -.-> ExtLedgerState) xs` argument, I think.

Reply:
A SanchNet sync would validate whether this works for the interesting multi-era case.

Reply:
> A SanchNet sync would validate whether this works for the interesting multi-era case.

Did so using a db-analyser pass with full block application. (FTR: Sanchonet is configured to hardfork into Alonzo at Genesis).

Also, this validated that the protocol version bug is still present in this PR (as expected, as this is not the job of this PR, it just enables a very simple fix).